File: frontend/src/types/scan.ts
// VulnScan Lite — TypeScript Type Definitions

export interface User {
  id: string;
  email: string;
  name?: string;
  created_at: string;
}

export interface TokenResponse {
  access_token: string;
  token_type: string;
  user?: User;
}

export type ScanStatus = 'queued' | 'running' | 'completed' | 'failed';

export type Grade = 'A' | 'B+' | 'B' | 'C' | 'D' | 'F';

export type CheckStatus = 'passed' | 'failed' | 'warning' | 'info';

export type FindingSeverity = 'critical' | 'high' | 'medium' | 'low' | 'info';

export type RiskLevel = 'Excellent' | 'Good' | 'Moderate' | 'Needs Improvement' | 'Poor';

export type AuthorizationType = 'user_owned' | 'organization_approved' | 'explicit_permission';

export interface ConsentAudit {
  consent_version: string;
  authorization_state: string;
  scan_status: string;
  confirmed_at?: string;
}

export interface ScanCreatePayload {
  url: string;
  authorization_type: AuthorizationType;
  target_confirmed: boolean;
  consent_version?: string;
  confirmed_ownership: boolean;
  confirmed_requests_acknowledged: boolean;
  confirmed_authorized_testing_only: boolean;
  confirmed_passive_analysis_understood: boolean;
  confirmed_responsibility_accepted: boolean;
}

export interface ScanCreateResponse {
  scan_id: string;
  status: ScanStatus;
  stage?: string;
  authorization_type?: string;
}

export interface ScanStatusResponse {
  scan_id: string;
  status: ScanStatus;
  stage?: string;
  message?: string;
  score?: number;
  grade?: Grade;
  risk_level?: RiskLevel | string;
  authorization_type?: string;
  error?: string;
}

export interface SecurityCheck {
  id: string;
  check_name: string;
  category: string;
  status: CheckStatus;
  points: number;
  description?: string;
}

export interface Finding {
  id: string;
  check_name: string;
  severity: FindingSeverity;
  category?: string;
  description: string;
  impact?: string;
  remediation?: string;
  affected_url?: string;
  evidence?: string;
  confidence?: string;
}

export interface SSLData {
  is_https: boolean;
  connection_successful: boolean;
  certificate_valid: boolean;
  certificate_expired: boolean;
  status?: string;
  subject?: string;
  issuer?: string;
  tls_version?: string;
  cipher?: string;
  not_before?: string;
  not_after?: string;
  days_until_expiry?: number;
  error?: string;
  points: number;
  description: string;
}

export interface HeaderCheck {
  header_name: string;
  present: boolean;
  value?: string;
  points: number;
  status: CheckStatus;
  description: string;
  remediation: string;
  category: string;
}

export interface HeaderData {
  checks: HeaderCheck[];
  raw_headers?: Record<string, string>;
  server?: string;
  x_powered_by?: string;
}

export interface CMSData {
  detected: boolean;
  cms_name?: string;
  version?: string;
  detection_source?: string;
  confidence: string;
  version_exposed: boolean;
  outdated_status?: string;
  description?: string;
}

export interface HTMLData {
  generator?: string;
  title?: string;
  description_meta?: string;
  technology_indicators: string[];
  form_count?: number;
  external_scripts?: number;
  has_https_links?: boolean;
  insecure_http_links?: string[];
  is_html?: boolean;
}

export interface ScanResult {
  scan_id: string;
  target_url: string;
  status: ScanStatus;
  stage?: string;
  authorization_type?: string;
  consent_audit?: ConsentAudit;
  score?: number;
  grade?: Grade;
  risk_level?: RiskLevel | string;
  duration_seconds?: number;
  started_at?: string;
  completed_at?: string;
  created_at: string;
  ssl_data?: SSLData;
  header_data?: HeaderData;
  cms_data?: CMSData;
  html_data?: HTMLData;
  security_checks: SecurityCheck[];
  findings: Finding[];
  error?: string;
}

export interface ScanHistoryItem {
  scan_id: string;
  target_url: string;
  status: ScanStatus;
  stage?: string;
  authorization_type?: string;
  score?: number;
  grade?: Grade;
  risk_level?: RiskLevel | string;
  findings_count?: number;
  created_at: string;
  completed_at?: string;
}

File: backend/app/schemas/scan.py
"""
VulnScan Lite — Scan Pydantic Schemas

Defines request and response schemas for scan operations, status polling,
authorization basis verification, structured consent enforcement, and reporting.
"""
from enum import Enum
from typing import Optional, List, Any, Dict
from pydantic import BaseModel, field_validator


class AuthorizationTypeEnum(str, Enum):
    USER_OWNED = "user_owned"
    ORGANIZATION_APPROVED = "organization_approved"
    EXPLICIT_PERMISSION = "explicit_permission"


CURRENT_CONSENT_VERSION = "Authorized Scanning Policy v1.0"


class ScanCreateRequest(BaseModel):
    """
    Request payload for POST /api/scan with mandatory authorization & consent.
    """
    url: str
    authorization_type: AuthorizationTypeEnum
    target_confirmed: bool = False
    consent_version: str = CURRENT_CONSENT_VERSION
    confirmed_ownership: bool = False
    confirmed_requests_acknowledged: bool = False
    confirmed_authorized_testing_only: bool = False
    confirmed_passive_analysis_understood: bool = False
    confirmed_responsibility_accepted: bool = False

    @field_validator("url")
    @classmethod
    def validate_url_syntax(cls, v: str) -> str:
        if not v or not isinstance(v, str):
            raise ValueError("URL cannot be empty.")
        v = v.strip()
        if not v.startswith(("http://", "https://")):
            raise ValueError("URL must start with http:// or https://")
        return v

    @field_validator("target_confirmed")
    @classmethod
    def validate_target_confirmation(cls, v: bool) -> bool:
        if not v:
            raise ValueError("Explicit target confirmation is required before scanning.")
        return v

    @field_validator("consent_version")
    @classmethod
    def validate_consent_version(cls, v: str) -> str:
        if v != CURRENT_CONSENT_VERSION:
            raise ValueError(
                f"Invalid or outdated consent version '{v}'. Current version is '{CURRENT_CONSENT_VERSION}'."
            )
        return v

    @field_validator(
        "confirmed_ownership",
        "confirmed_requests_acknowledged",
        "confirmed_authorized_testing_only",
        "confirmed_passive_analysis_understood",
        "confirmed_responsibility_accepted",
    )
    @classmethod
    def validate_all_consent_checkboxes(cls, v: bool, info) -> bool:
        if not v:
            field_name = info.field_name
            raise ValueError(f"Mandatory consent stipulation '{field_name}' must be explicitly confirmed.")
        return v


class ScanCreateResponse(BaseModel):
    """Response returned immediately after queuing a scan."""
    scan_id: str
    status: str = "queued"
    stage: str = "queued"
    authorization_type: Optional[str] = None


class ScanStatusResponse(BaseModel):
    """Lightweight response for status polling (GET /api/scan/{scan_id}/status)."""
    scan_id: str
    status: str                         # "queued" | "running" | "completed" | "failed"
    stage: Optional[str] = "queued"     # "queued" | "auth_verified" | "validating" | "policy_check" | "connecting" | "scanning" | "analyzing" | "generating_report" | "completed" | "failed"
    message: Optional[str] = None       # human-readable stage message
    score: Optional[float] = None
    grade: Optional[str] = None
    risk_level: Optional[str] = None
    authorization_type: Optional[str] = None
    error: Optional[str] = None


class SecurityCheckSchema(BaseModel):
    """Individual security check item."""
    id: str
    check_name: str
    category: str
    status: str                         # "passed" | "failed" | "warning" | "info"
    points: int
    description: Optional[str] = None

    model_config = {"from_attributes": True}


class FindingSchema(BaseModel):
    """Individual security finding item."""
    id: str
    check_name: str
    severity: str                       # "critical" | "high" | "medium" | "low" | "info"
    description: str
    remediation: Optional[str] = None
    impact: Optional[str] = None
    affected_url: Optional[str] = None
    evidence: Optional[str] = None
    confidence: Optional[str] = "high"

    model_config = {"from_attributes": True}


class ConsentAuditSchema(BaseModel):
    """Minimal public audit information for verified scan consent."""
    consent_version: str
    authorization_state: str
    scan_status: str
    confirmed_at: Optional[str] = None

    model_config = {"from_attributes": True}


class ScanResultResponse(BaseModel):
    """Detailed result payload for GET /api/scan/{scan_id}/result."""
    scan_id: str
    target_url: str
    status: str
    stage: Optional[str] = None
    authorization_type: Optional[str] = None
    consent_audit: Optional[ConsentAuditSchema] = None
    score: Optional[float] = None
    grade: Optional[str] = None
    risk_level: Optional[str] = None
    duration_seconds: Optional[float] = None
    started_at: Optional[str] = None
    completed_at: Optional[str] = None
    created_at: str
    ssl_data: Optional[Dict[str, Any]] = None
    header_data: Optional[Dict[str, Any]] = None
    cms_data: Optional[Dict[str, Any]] = None
    html_data: Optional[Dict[str, Any]] = None
    security_checks: List[SecurityCheckSchema] = []
    findings: List[FindingSchema] = []
    error: Optional[str] = None

    model_config = {"from_attributes": True}


class ScanHistoryItem(BaseModel):
    """Summary item for historical scan list."""
    scan_id: str
    target_url: str
    status: str
    stage: Optional[str] = None
    authorization_type: Optional[str] = None
    score: Optional[float] = None
    grade: Optional[str] = None
    risk_level: Optional[str] = None
    findings_count: int = 0
    created_at: str
    completed_at: Optional[str] = None

    model_config = {"from_attributes": True}

